File: csharp/Ray.cs
using System;
using System.Numerics;

namespace ToyRayTrace
{
    public readonly struct Ray
    {
        public Vector3 Origin { get; }
        public Vector3 Direction { get; }

        public Ray(in Vector3 origin, in Vector3 destination)
        {
            Origin = origin;
            Direction = Vector3.Normalize(destination);
        }

        //Origin + Direction * t;
        public Vector3 PointAtParameter(float t) => new Vector3(Origin.X + Direction.X * t, Origin.Y + Direction.Y * t, Origin.Z + Direction.Z * t);
    }
}

File: rust/src/raytrace/ray.rs
use super::vec3;
use super::Vec3;

#[derive(Default)]
pub struct Ray {
    pub origin: Vec3,
    pub direction: Vec3,
}

impl Ray {
    pub fn new(origin: Vec3, direction: Vec3) -> Self {
        Ray {
            origin,
            direction: vec3::normalize(direction),
        }
    }

    pub fn point_at_parameter(&self, t: f32) -> Vec3 {
        self.origin + self.direction * t
    }
}
